src/vm/eventtrace.cpp

```diff
@@ -4441,12 +4441,40 @@ extern "C"
 
         BOOLEAN bIsRundownTraceHandle = (context->RegistrationHandle==Microsoft_Windows_DotNETRuntimeRundownHandle);
 
-		// TypeSystemLog needs a notification when certain keywords are modified, so
-		// give it a hook here.
-		if (g_fEEStarted && !g_fEEShutDown && bIsPublicTraceHandle)
-		{
-			ETW::TypeSystemLog::OnKeywordsChanged();
-		}
+        if (g_fEEStarted && !g_fEEShutDown)
+        {
+            // TypeSystemLog needs a notification when certain keywords are modified, so
+            // give it a hook here.
+            if (bIsPublicTraceHandle)
+            {
+                ETW::TypeSystemLog::OnKeywordsChanged();
+            }
+
+            if (ControlCode == EVENT_CONTROL_CODE_ENABLE_PROVIDER || ControlCode == EVENT_CONTROL_CODE_DISABLE_PROVIDER)
```

Comment:
In order to support EventPipe, you will need this code to execute when the providers are enabled through EventPipe as well.  Otherwise, the GC won't get the message that the events have been enabled.  This can be done by adding a callback when the EventPipe provider is created, which occurs at https://github.com/dotnet/coreclr/blob/c1bbdae7964b19b7063074d36e6af960f0cdc3a0/src/scripts/genEventPipe.py#L141.

The callback that you've modified here is only used by ETW - not EventPipe.  Ideally it would get used by both, but I can tell you that there are things in this callback that EventPipe doesn't know how to handle, and so rather than you getting to feel that pain, I think you should do the following:

1. Add a new callback right next to this one with the same function signature.  You can call it something like EventPipeCallback.
2. Put your code in it.
3. Register the callback with EventPipe at the CreateProvider call in the Python script.
4. Call the new callback from within the existing callback that you've modified here so that you don't have to duplicate the code.

Then, as we find things here that we need to support in EventPipe, we can move them into the EventPipe callback which is a subset of the ETW callback.  Eventually, we'll get to the point where the ETW callback just calls the EventPipe callback and they merge back into one.

Reply:
will do - thanks!